File: IPWorks Samples/News Client/java/nntpread.java
/*
 * IPWorks 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworks.*;


public class nntpread extends ConsoleDemo {
	
	private static int lines;
	
	public static void main(String[] args) {
		
		if(args.length!=1) {
			
			System.out.println("usage: nntpread server ");
			System.out.println("");
			System.out.println("  server  the name or address of the NNTP server (e.g. nntp.aioe.org)");						
			System.out.println("\r\nExample: nntpread nntp.aioe.org");
			
		} else {
			
			Nntp nntp = new Nntp();

			int msgnum = 0; // current message number for next command

			try {
				
				nntp.addNntpEventListener(new DefaultNntpEventListener(){
					
					public void SSLServerAuthentication(NntpSSLServerAuthenticationEvent arg0) {
						arg0.accept=true; //this will trust all certificates and it is not recommended for production use	
					}							
					public void error(NntpErrorEvent e) {
						System.out.println("\nError " + e.errorCode + ": " + e.description);		
					}					
					public void groupList(NntpGroupListEvent e) {
						System.out.println(e.group);

						if (++lines == 23) {
							prompt("Press ENTER to continue", "...");
							lines = 0;
						}		
					}					
					public void groupOverview(NntpGroupOverviewEvent e) {
						System.out.println(e.articleNumber + "  " + e.subject);

						if (++lines == 23) {
							prompt("Press ENTER to continue", "...");
							lines = 0;
						}		
					}														
					public void header(NntpHeaderEvent e) {
						System.out.println(e.field + ": " + e.value);

						if (++lines == 23) {
							prompt("Press ENTER to continue", "...");
							lines = 0;
						}	
					}
					public void transfer(NntpTransferEvent e) {
						System.out.println(new String(e.text));

						if (e.EOL) {
							System.out.println();
							if (++lines == 23) {
								prompt("Press ENTER to continue", "...");
								lines = 0;
							}
						}	
					}
		        }); 
	    	
				nntp.setNewsServer(args[0]);
				nntp.connect();
				DisplayMenu();

				while (true) {
					String command = prompt("", ">");
					String[] argument = command.split("\\s");

					if (argument.length == 0 || argument[0].length() == 0) {
						// do nothing
					} else if (argument[0].equals("l")) {
						nntp.listGroups();
					} else if (argument[0].equals("s")) {
						if (argument.length > 1) {
							nntp.setCurrentGroup(argument[1]);
						} else {
							System.out.println("No newsgroup specified.");
						}
					} else if (argument[0].equals("v")) {
						msgnum = Integer.parseInt(argument[1]);
						nntp.setCurrentArticle(argument[1]);
						nntp.fetchArticle();
					} else if (argument[0].equals("n")) {
						msgnum++;
						nntp.setCurrentArticle(String.valueOf(msgnum));
						nntp.fetchArticle();
					} else if (argument[0].equals("h")) {
						nntp.setOverviewRange("0-");
						nntp.groupOverview();
					} else if (argument[0].equals("q")) {
						nntp.disconnect();
						break;					 
					} else {
						try {
							msgnum = Integer.parseInt(command);
							// allow user to enter the number of the message they want to view
							nntp.setCurrentArticle(command);
							nntp.fetchArticle();
						} catch (NumberFormatException e) {
							System.out.println("Bad command / Not implemented in demo.");
						}
					} // end of command checking
				} // end of main while loop
			} catch (Exception e) {
				displayError(e);
			}
		}
		
	}
	private static void DisplayMenu(){
		System.out.println("Readnews Commands");
		System.out.println("l                    list all available newsgroups");
		System.out.println("s <newsgroup>        select newsgroup");
		System.out.println("h                    give head lines of messages");
		System.out.println("v <message number>   view the content of selected message");
		System.out.println("n                    goto and view next message");		
		System.out.println("q                    quit, saving unresolved messages in mbox");
    }			
	
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksException) {
      System.out.print(" (" + ((IPWorksException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks Samples/News Client/cpp/nntpread.cpp
/*
 * IPWorks 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../../include/ipworks.h"

#include <ctype.h>

#define LINE_LEN 80


class MyNNTP: public NNTP
{
public:
	MyNNTP()
	{
		lines = 0;
	}
	virtual int FireGroupList(NNTPGroupListEventParams *e)
	{
		printf( "%s\n", e->Group );
		lines++;
		if ( lines == 22 )
		{
			printf("Press ENTER to continue...");
			getchar();
			lines = 0;
		}
		return 0;
	}

	virtual int FireGroupOverview(NNTPGroupOverviewEventParams *e)
	{
		printf( "%d  %s\n", e->ArticleNumber, e->Subject );
		lines++;
		if ( lines == 23 )
		{
			getchar();
			lines = 0;
		}
		return 0;
	}

	virtual int FireHeader(NNTPHeaderEventParams *e)
	{
		printf( "%s: %s\n", e->Field, e->Value );
		lines++;
		if ( lines == 23 )
		{
			getchar();
			lines = 0;
		}
		return 0;
	}

	virtual int FireTransfer(NNTPTransferEventParams *e)
	{
		printf( "%s\n", e->Text );
		lines++;
		if ( lines == 23 )
		{
			getchar();
			lines = 0;
		}
		return 0;
	}


	virtual int FireError(NNTPErrorEventParams *e)
	{
		printf("Error #%i: %s\n", e->ErrorCode, e->Description);
		exit(0);
	}

	virtual int FireSSLServerAuthentication(NNTPSSLServerAuthenticationEventParams *e)
	{
		if (e->Accept) return 0;
		printf("Server provided the following certificate:\nIssuer: %s\nSubject: %s\n",
			e->CertIssuer, e->CertSubject);
		printf("The following problems have been determined for this certificate: %s\n", e->Status);
		printf("Would you like to continue anyways? [y/n] ");
		if (getchar() == 'y') e->Accept = true;
		else exit(0);
		return 0;
	}
	int lines;
	char bkey;

};

int main(int argc, char * argv[])
{

	if (argc != 2) {

		fprintf(stderr, "usage: nntpread server\n");
		fprintf(stderr, "\n");
		fprintf(stderr, "  server  the name or address of the NNTP server\n");
		fprintf(stderr, "\nExample: nntpread nntp.aioe.org\n\n");
		printf("Press enter to continue.");
		getchar();

	}
	else{
		MyNNTP nntp;                // NNTP object
		int ret_code = 0;           // Return code:
		// = 0 when there is no error
		// = error code when there is an error

		char command[LINE_LEN];     // user's command
		char buffer[LINE_LEN];      // text buffer
		char *argument;             // arguments following command

		int msgnum = 0;             // current message number for next command

		if (ret_code = nntp.SetNewsServer(argv[1])) goto done;

		/*
		printf( "login\t: " );
		fgets(buffer,LINE_LEN,stdin);
		buffer[strlen(buffer)-1] = '\0';
		nntp.SetUser( buffer );

		printf( "password\t: " );
		fgets(buffer,LINE_LEN,stdin);
		buffer[strlen(buffer)-1] = '\0';
		nntp.SetPassword( buffer );
		*/
		if (ret_code = nntp.Connect()) goto done;

		printf("Type \"?\" for a list of commands.\n\n");
		while (1)
		{

			nntp.lines = 0;
			printf("> ");
			fgets(command, LINE_LEN, stdin);
			command[strlen(command) - 1] = '\0';
			argument = strtok(command, " \t\n");

			if (!strcmp(command, "?"))
			{

				printf("Readnews Commands\n"
					"l                               list all available newsgroups\n"
					"c <newsgroup>                   select newsgroup\n"
					"t <message number>              type messages\n"
					"n                               goto and type next message\n"
					"f                               give head lines of messages\n"
					"q                               quit, saving unresolved messages in mbox\n"
					"h                               print out active message headers\n");

			}

			else if (!strcmp(command, "c"))
			{

				if (argument = strtok(NULL, " \t\n"))
				{
					ret_code = nntp.SetCurrentGroup(argument);
				}
				else
				{
					printf("No newsgroup specified.\n");
				}

			}

			else if (!strcmp(command, "f"))
			{

				nntp.SetOverviewRange("0-");
				ret_code = nntp.GroupOverview();

			}

			else if (!strcmp(command, "h"))
			{

				nntp.SetOverviewRange("0-");
				ret_code = nntp.GroupOverview();

			}

			else if (!strcmp(command, "l"))
			{

				ret_code = nntp.ListGroups();

			}

			else if (!strcmp(command, "n"))
			{

				msgnum++;
				sprintf(buffer, "%i", msgnum);
				nntp.SetCurrentArticle(buffer);
				ret_code = nntp.FetchArticle();

			}

			else if (!strcmp(command, "q"))
			{

				ret_code = nntp.Disconnect();
				exit(0);

			}

			else if (!strcmp(command, "t"))
			{

				argument = strtok(NULL, " \t\n");
				msgnum = atoi(argument);
				nntp.SetCurrentArticle(argument);
				ret_code = nntp.FetchArticle();

			}

			else if (!strcmp(command, ""))
			{
				// do nothing
			}

			else
			{

				if (isdigit((int)command[0]))
				{
					// allow user to enter only the number
					//  of the message they want to view
					nntp.SetCurrentArticle(command);
					msgnum = atoi(command);
					ret_code = nntp.FetchArticle();
				}
				else
				{
					printf("Bad command / Not implemented in demo.\n");
				}

			} // end of command checking

			if (ret_code)     // Got an error.  The user is done.
			{
				printf("\nError: %d", ret_code);
				if (nntp.GetLastError())
				{
					printf(" \"%s\"\n", nntp.GetLastError());
				}
			}

			ret_code = 0;  // flush out the error

		}  // end of main while loop

	done:
		if (ret_code)     // Got an error.  The user is done.
		{
			printf("\nError: %d", ret_code);
			if (nntp.GetLastError())
			{
				printf(" \"%s\"\n", nntp.GetLastError());
			}
		}
		printf("Exiting... (press enter)\n");
		getchar();
		exit(ret_code);
		return 0;
	}
}


